clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -298,63 +357,43 @@ protected void onJoinComplete(int generation,
 
         // execute the user's callback after rebalance
-        ConsumerRebalanceListener listener = subscriptions.rebalanceListener();
-
+        final AtomicReference<Exception> firstException = new AtomicReference<>(null);
+        Set<TopicPartition> addedPartitions = new HashSet<>(assignedPartitions);
+        addedPartitions.removeAll(ownedPartitions);
         switch (protocol) {
             case EAGER:
-                if (!ownedPartitions.isEmpty()) {
-                    log.info("Coordinator has owned partitions {} that are not revoked with {} protocol, " +
-                        "it is likely client is woken up before a previous pending rebalance completes its callback", ownedPartitions, protocol);
-                }
+                // assign partitions that are not yet owned
+                firstException.compareAndSet(null, maybeInvokePartitionsAssigned(addedPartitions));
 
-                log.info("Setting newly assigned partitions: {}", Utils.join(assignedPartitions, ", "));
-                try {
-                    listener.onPartitionsAssigned(assignedPartitions);
-                } catch (WakeupException | InterruptException e) {
-                    throw e;
-                } catch (Exception e) {
-                    log.error("User provided listener {} failed on partition assignment", listener.getClass().getName(), e);
-                }
                 break;
 
             case COOPERATIVE:
-                assignAndRevoke(listener, assignedPartitions, ownedPartitions);
+                Set<TopicPartition> revokedPartitions = new HashSet<>(ownedPartitions);
+                revokedPartitions.removeAll(assignedPartitions);
 
-                break;
-        }
+                log.info("Updating with newly assigned partitions: {}, compare with already owned partitions: {}, " +
+                        "newly added partitions: {}, revoking partitions: {}",
+                    Utils.join(assignedPartitions, ", "),
+                    Utils.join(ownedPartitions, ", "),
+                    Utils.join(addedPartitions, ", "),
+                    Utils.join(revokedPartitions, ", "));
 
-    }
+                // add partitions that were not previously owned but are now assigned
+                firstException.compareAndSet(null, maybeInvokePartitionsAssigned(addedPartitions));
 
-    private void assignAndRevoke(final ConsumerRebalanceListener listener,
-                                 final Set<TopicPartition> assignedPartitions,
-                                 final Set<TopicPartition> ownedPartitions) {
-        Set<TopicPartition> addedPartitions = new HashSet<>(assignedPartitions);
-        Set<TopicPartition> revokedPartitions = new HashSet<>(ownedPartitions);
-        addedPartitions.removeAll(ownedPartitions);
-        revokedPartitions.removeAll(assignedPartitions);
+                // revoked partitions that was previously owned but no longer assigned
```

Comment:
nit: **revoke** partitions that **were** previously owned but **are** no longer assigned